python_ta/typecheck/base.py

```diff
@@ -52,33 +99,70 @@ class TypeFail(TypeResult):
     """Represents the result of a failed type check operation
     Contains error message
     """
-
-    def __init__(self, tnode1: _TNode = None, tnode2: _TNode = None, src_node: NodeNG = None):
-        if isinstance(tnode1, str):
-            self.msg = tnode1
-            self.tnode1 = None
-        else:
-            self.tnode1 = tnode1
-            self.msg = ""
-        self.tnode2 = tnode2
-        self.src_node = src_node
+    def __init__(self, msg: str):
+        self.msg = msg
         super().__init__(self.msg)
 
     def __str__(self):
-        if self.tnode1 is None:
-            return f'TypeFail: {self.msg}'
-        if self.src_node:
-            return 'TypeFail: %s <-> %s at %s' % \
-                   (self.tnode1.type, self.tnode2.type, self.src_node.as_string())
-        else:
-            return 'TypeFail: %s <-> %s' % (self.tnode1.type, self.tnode2.type)
+        return f'TypeFail: {self.msg}'
 
     def bind(self, _):
         return self
 
     def add_msg(self, msg: str):
         self.msg = msg
 
+    def get_reasons(self):
+        """ Returns list of reasons for type inference failure """
+        raise NotImplementedError
+
+
+class TypeFailUnify(TypeFail):
+    """
+    TypeFailUnify occurs when two types fail to unify.
+    Should be initialized with two valid _TNodes, and the astroid node where the failure occurred
+    """
+    def __init__(self, tnode1: _TNode, tnode2: _TNode, src_node: NodeNG = None):
+        if isinstance(tnode1, _TNode) and isinstance(tnode2, _TNode):
+            self.tnode1 = tnode1
```

Comment:
Make an attribute storing a set of tnodes. You can describe a representation invariant that says this set must always have size 2, if you want.